File: kinv/adapters/git.py
"""Which files git already has uncommitted changes to.

Not to police your working tree: after a write, ``git diff`` should show *only*
what the tool did. A directory that is not a repository, or a machine with no
git, returns nothing — the guard exists where it can help and is silent where
it cannot.
"""

from __future__ import annotations

import os
import subprocess


def dirty_files(files: list[str]) -> list[str]:
    if not files:
        return []
    cwd = os.path.dirname(files[0]) or "."
    try:
        creation = subprocess.CREATE_NO_WINDOW if os.name == "nt" else 0
        result = subprocess.run(
            ["git", "status", "--porcelain", "--", *files],
            cwd=cwd,
            # Never the server's own stdin: on Windows, git's runtime probes the
            # handle it inherits, and a pipe with a read pending on it blocks the
            # probe — which hung every schematic write in the browser tests.
            stdin=subprocess.DEVNULL,
            capture_output=True,
            creationflags=creation,
            check=True,
        )
    except (OSError, subprocess.CalledProcessError):
        return []

    changed = set()
    for line in result.stdout.decode("utf8", errors="replace").split("\n"):
        entry = line[3:].strip()
        if entry == "":
            continue
        # `XY path`, and ` -> ` for a rename; the path is what matters here.
        if " -> " in entry:
            entry = entry.split(" -> ")[1]
        changed.add(os.path.normcase(os.path.abspath(os.path.join(cwd, entry.strip('"')))))
    return [f for f in files if os.path.normcase(os.path.abspath(f)) in changed]

File: src/adapters/git.ts
import { execFileSync } from "node:child_process";
import { dirname, resolve } from "node:path";

/**
 * Which of these files git already has uncommitted changes to.
 *
 * The point is not to police your working tree: it is that after this write
 * runs, `git diff` should show *only* what the tool did. A file that was
 * already modified turns "here is what changed" into a guessing game, and the
 * `.bak` beside it is a worse answer than a commit.
 *
 * A directory that is not a repository, or a machine with no git, returns
 * nothing: the guard exists where it can help and is silent where it cannot.
 */
export function dirtyFiles(files: readonly string[]): string[] {
  if (files.length === 0) return [];
  const cwd = dirname(files[0] as string);
  try {
    const out = execFileSync("git", ["status", "--porcelain", "--", ...files], {
      cwd,
      encoding: "utf8",
      stdio: ["ignore", "pipe", "ignore"],
      windowsHide: true,
    });
    // `XY path`, and ` -> ` for a rename; the path is what matters here.
    const changed = out
      .split("\n")
      .map((line) => line.slice(3).trim())
      .filter((line) => line !== "")
      .map((line) => (line.includes(" -> ") ? (line.split(" -> ")[1] as string) : line))
      .map((line) => resolve(cwd, line.replace(/^"|"$/g, "")));

    return files.filter((f) => changed.includes(resolve(f)));
  } catch {
    return [];
  }
}
